pkgs/crypto/lib/src/hash_sink.dart

```diff
@@ -79,18 +78,48 @@ abstract class HashSink implements Sink<List<int>> {
   void add(List<int> data) {
     if (_isClosed) throw StateError('Hash.add() called after close().');
     _lengthInBytes += data.length;
-    _pendingData.addAll(data);
-    _iterate();
+    _addData(data);
+  }
+
+  void _addData(List<int> data) {
+    var dataIdx = 0;
+    var currentChunkNextIndex = _currentChunkNextIndex;
+    final size = _currentChunk.length;
+    _byteDataView ??= _currentChunk.buffer.asByteData();
+    while (true) {
+      // Is there enough data left in [data] for a full chunk?
+      var restEnd = currentChunkNextIndex + data.length - dataIdx;
+      if (restEnd < size) {
+        // No. Just add into [_currentChunk].
+        _currentChunk.setRange(currentChunkNextIndex, restEnd, data, dataIdx);
+        _currentChunkNextIndex = restEnd;
+        return;
+      }
+
+      // Yes. Fill out [_currentChunk] and process it.
+      _currentChunk.setRange(currentChunkNextIndex, size, data, dataIdx);
+      dataIdx += size - currentChunkNextIndex;
+
+      // Now do endian conversion to words.
+      var j = 0;
```

Comment:
this can't be done in a "normal" for loop?

Reply:
It can. But this is faster.